sw/device/lib/drivers/rv_plic/rv_plic.c

```diff
@@ -34,337 +81,417 @@ typedef struct plic_reg_info {
+
+/****************************************************************************/
+/**                                                                        **/
+/*                            GLOBAL VARIABLES                              */
+/**                                                                        **/
+/****************************************************************************/
+
+int8_t external_intr_flag = 0;
+
+/****************************************************************************/
+/**                                                                        **/
+/*                           EXPORTED FUNCTIONS                             */
+/**                                                                        **/
+/****************************************************************************/
+
+__attribute__((weak)) void handler_irq_uart(void) {
+  printf("WEAK UART IRQ triggered!\n");
+  volatile int i=0;
+  i++;
 }
 
-/**
- * Get the address of the first target N interrupt enable register (IEN0).
- */
-static ptrdiff_t plic_irq_enable_base_for_target(dif_plic_target_t target) {
-  ptrdiff_t range = RV_PLIC_IE0_MULTIREG_COUNT * sizeof(uint32_t);
-  return RV_PLIC_IE0_0_REG_OFFSET + (range * target);
+__attribute__((weak)) void handler_irq_gpio(void) {
+  printf("WEAK GPIO IRQ triggered!\n");
+  volatile int i=0;
+  i++;
 }
 
-/**
- * Get the address of the first target N software interrupt register (MSIPN).
- */
-static ptrdiff_t plic_software_irq_base_for_target(dif_plic_target_t target) {
-  return RV_PLIC_MSIP0_REG_OFFSET + (target * sizeof(uint32_t));
+__attribute__((weak)) void handler_irq_i2c(void) {
+  printf("WEAK I2C IRQ triggered!\n");
+  volatile int i=0;
+  i++;
 }
 
-/**
- * Get the address of the first target N threshold register (THRESHOLDN).
- */
-static ptrdiff_t plic_threshold_base_for_target(dif_plic_target_t target) {
-  return RV_PLIC_THRESHOLD0_REG_OFFSET + (target * sizeof(uint32_t));
+__attribute__((weak)) void handler_irq_spi(void) {
+  printf("WEAK SPI IRQ triggered!\n");
+  volatile int i=0;
+  i++;
 }
 
-/**
- * Get the address of the first target N claim complete register (CCN).
- */
-static ptrdiff_t plic_claim_complete_base_for_target(dif_plic_target_t target) {
-  return RV_PLIC_CC0_REG_OFFSET + (target * sizeof(uint32_t));
-}
 
 /**
- * Get a target and an IRQ source specific Interrupt Enable register info.
- */
-static plic_reg_info_t plic_irq_enable_reg_info(dif_plic_irq_id_t irq,
-                                                dif_plic_target_t target) {
-  ptrdiff_t offset = plic_offset_from_reg0(irq);
-  return (plic_reg_info_t){
-      .offset = plic_irq_enable_base_for_target(target) + offset,
-      .bit_index = plic_irq_bit_index(irq),
-  };
+ * Returns the irq_sources_t source type of a given irq source ID
+*/
+static irq_sources_t plic_get_irq_src_type(plic_irq_id_t irq_id)
+{
+  if(irq_id >= UART_ID_START && irq_id <= UART_ID_END)
```

Comment:
Try to put always negative-first implementations in multi-branch code, by doing so you can avoid one comparison per if

Reply:
also, i think we can optimize this by using function listing